op.py

```diff
@@ -149,6 +153,7 @@ def check_file_exists(value):
                                  'stop',
                                  'teardown',
                                  'create_users',
+                                 'addon_istio',
```

Comment:
Here's a thought that I'd like your opinion on:

Instead of adding a new action called `addon_istio`, let's make it more generic and call it `addon`. Then we can add additional addons and keep the same command structure.

This require a bit of a re-write on how `argparse` is used here and would need to use `subparsers`. I've taken the time to convert it below (not fully tested yet) but should solve the problem.

```python
if __name__ == '__main__':
    parser = argparse.ArgumentParser()
    subparsers = parser.add_subparsers()

    parser_provision = subparsers.add_parser('provision')
    parser_provision.set_defaults(action='provision')

    parser_start = subparsers.add_parser('start')
    parser_start.set_defaults(action='start')

    parser_stop = subparsers.add_parser('stop')
    parser_stop.set_defaults(action='stop')

    parser_teardown = subparsers.add_parser('teardown')
    parser_teardown.set_defaults(action='teardown')

    parser_create_users = subparsers.add_parser('create_users')
    parser_create_users.set_defaults(action='create_users')

    parser_shell = subparsers.add_parser('shell')
    parser_shell.set_defaults(action='shell')

    parser_ssh = subparsers.add_parser('ssh')
    parser_ssh.set_defaults(action='ssh')

    parser_addon = subparsers.add_parser('addon')
    parser_addon.set_defaults(action='addon')
    parser_addon.add_argument('addon',
                              choices=['istio',])

    parser.add_argument('--env-file',
                        required=True,
                        type=check_file_exists,
                        help='file of environment variables')
    parser.add_argument('--vars-file',
                        required=True,
                        type=check_file_exists,
                        help='file of ansible variables')
    parser.add_argument('--no-update',
                        action='store_true')
    parser.add_argument('--dev',
                        action='store_true')
    known_args, extra_args = parser.parse_known_args()

    if os.geteuid() != 0:
        print('This script requires root privileges.')
        exit(1)

    try:
        op = OpenShiftProvision(known_args.env_file,
                                known_args.vars_file,
                                known_args.no_update,
                                known_args.dev,
                                extra_args)
    except ContainerRuntimeMissingError:
        print('\n'.join([
            'You do not have a supported container runtime installed.',
            '',
            'This script supports the following container runtimes:',
            '\n'.join('  - {}'.format(i) for i in SUPPORTED_CONTAINER_RUNTIMES),
            '',
            'Please install one of those options and try again.'
        ]))

    if known_args.action == 'provision':
        op.provision()
    elif known_args.action == 'start':
        op.start_instances()
    elif known_args.action == 'stop':
        op.stop_instances()
    elif known_args.action == 'teardown':
        op.teardown()
    elif known_args.action == 'create_users':
        op.create_users()
    elif known_args.action == 'addon':
        if known_args.addon == 'istio':
            op.addon_istio()
    elif known_args.action == 'shell':
        op.shell()
    elif known_args.action == 'ssh':
        op.ssh()
```

Let me know what you think. If you like it, then I can create a PR for your branch and then we can update this PR with the new code.

Reply:
Great idea, I thought about doing exactly this but then backed away because it'd take a little more time and I wasn't sure if we'd need more addons or not... knative maybe or 3 scale would be cool.  But then again things in 4.x will be a lot different and I'm guessing this provisioner will be a 3.x forever tool.

Reply:
I'm thinking even other tools (e.g. Gitlab, Gogs, etc.) could be addons to help with workshop provsioning.

My intention is to update this to also do 4.x as well. Contrary to popular belief, running `openshift-install` to install 4.x is only the first step. You then have to configure the cluster to be usable for what you need. In 3.x, those decisions had to be made at install time and are part of the inventory, but with 4.x you will have to do those after the cluster is up and running (e.g. deploy certs, deploy OCS, configure authentication, add users, etc.) So the code base for 4.x won't have much overlap with what's here for 3.x, but I still intend to keep the format the same as much as I can so that when you specify a 4.x version it'll just do the right thing for you.